subsys/llext/llext_load.c

```diff
@@ -539,6 +526,37 @@ static int llext_copy_symbols(struct llext_loader *ldr, struct llext *ext,
 	return 0;
 }
 
+/*
+ * Free resources only used during loading. Also, in case an error was
+ * detected, free everything that was allocated.
+ */
+static void llext_load_cleanup(struct llext_loader *ldr, struct llext *ext, int ret)
+{
+	/* free resources only used during loading */
+	llext_free(ldr->sect_map);
+	ldr->sect_map = NULL;
+
+	if (ldr->sect_hdrs_on_heap) {
+		llext_free(ldr->sect_hdrs);
+	}
+	ldr->sect_hdrs = NULL;
+
+	if (!IS_ENABLED(CONFIG_LLEXT_LOG_LEVEL_DBG)) {
+		/* keep the computed symbol tables for debugging */
+		llext_free(ext->sym_tab.syms);
+		ext->sym_tab.sym_cnt = 0;
+		ext->sym_tab.syms = NULL;
+	}
+
+	if (ret != 0) {
+		/* free everything */
+		llext_free_sections(ext);
+		llext_free(ext->exp_tab.syms);
+		ext->exp_tab.sym_cnt = 0;
+		ext->exp_tab.syms = NULL;
+	}
```

Comment:
Nit: I would regroup symbol table freeing under a single `if (ret != 0 || !IS_ENABLED(CONFIG_LLEXT_LOG_LEVEL_DBG))`.
In current state, I find it difficult to see the difference between this block and the one above (this one also calls `llext_free_section`).

Reply:
I know it's difficult to catch `sym_tab` (ELF symbols, freed unless in debug mode) vs `exp_tab` (llext public symbols, freed in case of error). I will definitely improve comments if I need to do a second iteration: the debug one should have been outside the `if` :man_facepalming: